### PR TITLE
Issue15 - Testing mtcpclient part1 (stddev function)

### PR DESCRIPTION
**Pull request by dcaba**
First commit with a minimum contribution to the unit tests: 
https://github.com/dachad/tcpgoon/issues/15

Applying TDT for the stddev calculations. Some minor changes/additions introduced in the tested code, but some others expected (see TODO) once we actually increase coverage. 

Added the race flag to our scripted go test execution.

As usual, I will squash and merge once approved

### CODE REVIEW

File: _script/test [hunk 1]
```diff
@@ -6,4 +6,4 @@ cd $(dirname $0)/..
 
 echo "[*] Running Tests"
 
-go test -v -cover ./...
+go test -v -cover -race ./...
```
Comment (chadell) [inline _script/test]: 👏  Nice! I've read https://blog.golang.org/race-detector and it's worthy to have it

File: tcpclient/connection.go [hunk 2]
```diff
@@ -27,6 +27,20 @@ const (
 	ConnectionError        ConnectionStatus = 4
 )
 
+// NewConnection initializes a connection given all values that are actually stored internally. This is just being used
```
Comment (chadell) [inline tcpclient/connection.go]: You could use it also in the connectionManagement...
Btw, this NewConnection is not really a NewConnection which would have a default status and metrics=0, I 'd says it's more a SetConnection 🤔

File: mtcpclient/calculate.go [hunk 3]
```diff
@@ -39,13 +45,19 @@ func (gc GroupOfConnections) calculateMetricsReport(status tcpclient.ConnectionS
...
 	var sd float64
-
 	for _, item := range gc {
 		if item.GetConnectionStatus() == status {
+			nItems++
 			sd += math.Pow(float64(item.GetTCPProcessingDuration(status))-float64(mr.avg), 2)
 		}
 	}
-	return time.Duration(math.Sqrt(sd / float64(time.Duration(mr.numberOfConnections))))
+	if nItems == 0 {
```
Comment (chadell) [inline mtcpclient/calculate.go]: did you find this while testing it? 😃

File: mtcpclient/calculate.go [hunk 4]
```diff
@@ -39,13 +45,19 @@ func (gc GroupOfConnections) calculateMetricsReport(status tcpclient.ConnectionS
 }
 
 func (gc GroupOfConnections) calculateStdDev(status tcpclient.ConnectionStatus, mr metricsCollectionStats) time.Duration {
+	// TODO: passing the whole mr struct looks overkilling, given we only want a single value, the average, and maybe
```
Comment (chadell) [inline mtcpclient/calculate.go]: I agree, you only need the average, but I was also using the numberOfConnections that you have replaced by nItems and recalculate again within the loop

File: mtcpclient/calculate.go (hunk 3 above)
Comment (chadell) [reply mtcpclient/calculate.go]: why nItems instead of numberOfConnections?

File: mtcpclient/calculate.go [hunk 5]
```diff
@@ -17,6 +17,12 @@ type metricsCollectionStats struct {
 }
 
 func (gc GroupOfConnections) calculateMetricsReport(status tcpclient.ConnectionStatus) (mr metricsCollectionStats) {
+	// TODO: There's something i don't like... initiatlizing values in the first loop, and the standard deviation
```
Comment (chadell) [inline mtcpclient/calculate.go]: 👍

File: mtcpclient/calculate.go [hunk 6]
```diff
@@ -17,6 +17,12 @@ type metricsCollectionStats struct {
...
+	//  requiring an extra pass considering all items... i'd move initialization out of the loop, and maybe iterate
+	//  over a filtered list rather than several loops over the original one, and maybe use specific generic functions...
+	//  requires further thinking in any case
+	// TODO: (2) rather than calculate a metrics report with a filter on the connection status, lets just promote splitting
```
Comment (chadell) [inline mtcpclient/calculate.go]: It's an option, but is this more efficient? I mean you are doing a first loop to filter, store a new structure with the filtered list and going through it to get the metrics.

File: _script/test (hunk 1 above)
Comment (dcaba) [reply _script/test]: :+1:

File: tcpclient/connection.go (hunk 2 above)
Comment (dcaba) [reply tcpclient/connection.go]: Well, here we just adopt the standard naming convention: a constructor should follow the pattern New$ObjectName , and this is a very verbose constructor, not pretending to replace the standard instance creation (that could have the same name but less parameters, as yes, a new connection should have some defaults we are allowing the tests to override in this one)

File: mtcpclient/calculate.go (hunk 3 above)
Comment (dcaba) [reply mtcpclient/calculate.go]: I want to decouple the stdDev function as much as possible from our structures and "external requirements": i really feel making it autonomous will bring clarity...

File: mtcpclient/calculate.go (hunk 4 above)
Comment (dcaba) [reply mtcpclient/calculate.go]: See previous comment: making functions as autonomous as possible is a good principle from my point of view. As well as making dependencies as explicit as possible...

File: mtcpclient/calculate.go (hunk 3 above)
Comment (chadell) [reply mtcpclient/calculate.go]: 👍

File: mtcpclient/calculate.go (hunk 4 above)
Comment (chadell) [reply mtcpclient/calculate.go]: yes, I also agree

File: mtcpclient/calculate.go (hunk 6 above)
Comment (dcaba) [reply mtcpclient/calculate.go]: Well, if all decisions are made based on efficiency, probably we should be programming in assembler... of course, you can try to reduce "iterations" and do everything you need on each item... but this probably means doing too much staff in the same function, and this is a very bad practice unless, as you point, is critical in terms of performance. In fact, we've already broken this "practice" by moving std deviation outside of this function (and assuming more than one iteration then...). Making an specific assessment for this case... a reporting function that runs at the very end of the program (not part of the "main" program functionality, that is doing the tcp conn mgmt), that works with an structure that won't have more than 1k-10k elements... do you really think is worth to sacrifice readability / simplicity in favour of avoiding an extra loop over this structure? In fact, making a first loop to pick the elements that are actually important will minimise the number of items you will consider in future loops... So my bet on removing methods tight to a "group of connections" that use filters + do the logic against the filtered items, versus methods that do filtering and different methods that work against a group of connections (that probably have been already filtered) is extremely clear and I'm pretty sure you will notice the difference from a "clean-code" perspective once you see this implemented :)

Comment (chadell) [reply mtcpclient/calculate.go]: Yup, I agree with you... we should code in assembler (just kidding)
I agree with you that cleanliness is more important writing code than performance (if it's not critical), and splitting responsibilities makes everything easier to maintain an evolve.
I was just wondering when to take the step in each direction...

**Comment by chadell** [comment]
@dcaba , do you want to merge this PR as a first step on mtcpconnection testing? or are you expecting to add more commits to it? (just to understand if you are waiting for my approve)

**Comment by dcaba** [comment]
Hi,

  I'd merge to keep prs as small as possible, and benefit from them soon ;)

**Pull request merged**